app/api/dds/v1/trashbin_api.rb

```diff
@@ -0,0 +1,129 @@
+module DDS
+  module V1
+    class TrashbinAPI < Grape::API
+      desc 'View Trashbin Item details' do
+        detail 'Show Details of a Trashbin Item.'
+        named 'show trashbin item'
+        failure [
+          [200, 'Success'],
+          [401, 'Unauthorized'],
+          [403, 'Forbidden'],
+          [404, 'Item Does not Exist'],
+          [404, 'Object kind not supported']
+        ]
+      end
+      params do
+        requires :object_kind, type: String, desc: 'Object Kind'
+        requires :object_id, type: String, desc: 'Object UUID'
+      end
+      get '/trashbin/:object_kind/:object_id', root: false do
+        authenticate!
+        object_kind = KindnessFactory.by_kind(params[:object_kind])
+        trashed_object = object_kind.find_by!(id: params[:object_id], is_deleted: true, is_purged: false)
+        authorize trashed_object, :show?
+        trashed_object
+      end
+
+      desc 'Restore a Trashbin Item' do
+        detail 'Restores the item, and any children, to an undeleted status to the specified parent folder or project.'
+        named 'restore trashbin item'
+        failure [
+          [200, 'Success'],
+          [404, 'Parent object does not exist or is itself in the trashbin'],
+          [401, 'Unauthorized'],
+          [403, 'Forbidden'],
+          [404, 'Object or Parent kind not supported'],
+          [404, 'Object not found in trash bin'],
+          [404, 'Parent object does not exist or is itself in the trashbin']
+        ]
+      end
+      params do
+        requires :object_kind, type: String, desc: 'Object Kind'
+        requires :object_id, type: String, desc: 'Object UUID'
+        optional :parent, desc: "Parent. Required unless restoring a FileVersion", type: Hash do
+          optional :kind, type: String, desc: 'Parent Kind'
+          optional :id, type: String, desc: 'Parent UUID'
+        end
+      end
+      put '/trashbin/:object_kind/:object_id/restore', root: false do
+        authenticate!
+        parent_params = declared(params, {include_missing: false}, [:parent])
+        object_kind = KindnessFactory.by_kind(params[:object_kind])
+        purge_object = object_kind.find_by!(id: params[:object_id], is_deleted: true)
+        unless purge_object.class.include? Restorable
+          error_json = {
+            "error" => "404",
+            "code" => "not_provided",
+            "reason" => "#{purge_object.class.to_s} Not Found",
+            "suggestion" => "#{purge_object.kind} is not Restorable"
+          }
+          error!(error_json, 404)
+        end
+        unless purge_object.is_purged?
```

Comment:
Nothing happens if `purge_object.is_purged?` is true? Seems like an immutable validation error or something should result from this...